### ISSUE TITLE
[C][Linux] Cross compilation for a raspberry (--use-websockets) fails

### ISSUE THREAD

**Issue by lmussier** [Azure/azure-iot-sdks#525]
Hi,

I've follow the steps described [here](https://github.com/Azure/azure-iot-sdks/blob/develop/c/doc/SDK_cross_compile_example.md) and successfully cross compiled the SDK.

Since I'm interested in AMQP/websockets I've try the following.
````
./build.sh --toolchain-file /home/lmussier/workspace/Agents/iothub/agent/cmake/Pi.cmake --use-websockets -cl -D__STDC_NO_ATOMICS__
```

And get this error
```
/home/lmussier/workspace/official/c/azure-c-shared-utility/adapters/uniqueid_linux.c:7:23: fatal error: uuid/uuid.h: No such file or directory
 #include <uuid/uuid.h>
                       ^
compilation terminated.
azure-c-shared-utility/CMakeFiles/aziotsharedutil.dir/build.make:782: recipe for target 'azure-c-shared-utility/CMakeFiles/aziotsharedutil.dir/adapters/uniqueid_linux.c.o' failed
make[2]: *** [azure-c-shared-utility/CMakeFiles/aziotsharedutil.dir/adapters/uniqueid_linux.c.o] Error 1
CMakeFiles/Makefile2:986: recipe for target 'azure-c-shared-utility/CMakeFiles/aziotsharedutil.dir/all' failed
make[1]: *** [azure-c-shared-utility/CMakeFiles/aziotsharedutil.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

After inspecting the CMakes, i dont see how the build can work in a cross compilation setup.
[pkg_search_module](https://github.com/Azure/azure-c-shared-utility/blob/develop/tests/uniqueid_unittests/CMakeLists.txt#L33) is to be for searching via pkg-config, it won't work here.

I think there is also problems when CMakes are written like this

`target_link_libraries(mqtt_client_sample ssl crypto)`

The generated build won't find the libs. I think I can make the build (at least for AMQP/WS) works for the Raspberry.
Do you have already initiate something by you side ?

**Comment by markrad** [comment]
Hi @lmussier,

I put together the cross compile environment as described in the document you referenced and ran the command 
```
./build.sh --toolchain-file toolchain-rpi.cmake --skip-e2e-tests --use-websockets --skip-unittests -cl -D__STDC_NO_ATOMICS__ -cl --sysroot=$RPI_ROOT
```
I encountered one issue caused by the libwebsockets CMakeLists.txt specifying the -Werror flag which will cause warnings to be treated as errors. This then failed on the warning about a redefinition of \__STDC_NO_ATOMICS__. Once I removed that gcc directive from the CMakeLists.txt I was able to successfully build the SDK. Possibly the issue you are seeing has already been addressed. Can you update to the latest version and rerun your test please? 

Thank you.

_Mark Radbourne MSFT_

**Comment by lmussier** [comment]
You are right the issue with uuid is no more.
Unset -werror in libwebsocket make the compilation possible, I think we have to find a solution to avoid editing third party code.

Thanks for the update.

**Comment by markrad** [comment]
Hi @lmussier,

Thank you for your response. I will propose that the -Werror flag is removed. It's not used anywhere else in the SDK build so I don't think that will be an issue.

_Mark Radbourne MSFT_

**Comment by lmussier** [comment]
Sounds fair to me.
I'll let you close the issue when you are done, thanks !

2016-05-14 2:08 GMT+02:00 markrad <notifications@github.com>:

**Comment by markrad** [comment]
Hi @lmussier,

I revisited this issue and find that the libwebsockets GitHub project is not actually a MSFT project. The only way to get the -Werror removed would be to issue a pull request to this other project. I don't know how successful that would be. Perhaps you be willing to raise the issue with them?

https://github.com/warmcat/libwebsockets

_Mark Radbourne MSFT_

**Comment by lmussier** [comment]
@markrad I bet the -Werror won't be removed from the libwebsocket, this is part of the lib compile flags. libwebsocket compiles fine itself, the flag we add while compiling for the raspberry should not hit libwesocket :s

**Comment by markrad** [comment]
I ran it by them with the response being as you predicted. I'll dig around in the build system some more to see if I can find some way to circumvent the issue.

**Comment by markrad** [comment]
Hi @lmussier,

So I'm still not entirely sure why the definition of \__STDC_NO_ATOMICS__ is considered redefined. My best guess is that the build system is somehow adding the define to the command line twice. However, rather then spend too much time on that just yet I did find an easy circumvention that does not require any code modification. When you run build.sh for a cross compile add the additional flag _-cl -Wno-error_. For example your invocation would look something like this:
```
./build.sh --toolchain-file toolchain-rpi.cmake --skip-unittests -cl -D__STDC_NO_ATOMICS__ -cl --sysroot=$RPI_ROOT -cl -Wno-error --use-websockets
```
This will cause the compiler to ignore the -Werror flag and the build will succeed. 
Since there is a simple circumvention, I have closed this issue. Thank you for bring this to our attention.

_Mark Radbourne MSFT_
